### PR TITLE
Let bgp docker generate configuration by itself

### PR DESCRIPTION
**Pull request by taoyl-ms**
A temp solution and call for comments.

Ideally python scripts and config.py should be put into docker-base or another layer of base image as the same logic will be shared by different dockers

### CODE REVIEW

File: dockers/docker-bgp/Dockerfile [hunk 1]
```diff
@@ -3,11 +3,15 @@ FROM docker-base
 COPY deps/quagga_*.deb /deps/
 RUN dpkg_apt() { [ -f $1 ] && { dpkg -i $1 || apt-get -y install -f; } || return 1; } &&        \
     dpkg_apt /deps/quagga_*.deb &&                                                              \
+    apt-get update &&                                                                           \
+    apt-get -y install -f python-jinja2 python-netaddr python-ipaddr python-lxml &&             \
```
Comment (lguohan) [inline dockers/docker-bgp/Dockerfile]: move them to docker base as all the docker are going to use them.

File: dockers/docker-bgp/bgpd.conf.j2 [hunk 2]
```diff
@@ -0,0 +1,63 @@
+!
+{% block banner %}
+! =========== Managed by Ansible DO NOT EDIT! ========================
+! generated by templates/quagga/bgpd.conf.j2 using minigraph_facts.py
+! file: bgpd.conf
+!
```
Comment (qiluo-msft) [inline dockers/docker-bgp/bgpd.conf.j2]: Not true

File: dockers/docker-bgp/Dockerfile [hunk 3]
```diff
@@ -3,11 +3,15 @@ FROM docker-base
...
     apt-get clean -y && apt-get autoclean -y && apt-get autoremove -y &&                        \
     rm -rf /deps
 
 COPY daemons /etc/quagga/
+COPY ["*.j2", "config.sh", "template_list", "*.py", "/etc/swss/"]
```
Comment (lguohan) [inline dockers/docker-bgp/Dockerfile]: can we embed the mountpoint here, e.g.,

VOLUME ['/etc/sonic']

Then, in the base image, we can put minigraph in the /etc/sonic/,

also, the VOLUME is in baseimage as it is shared by all dockers.

File: dockers/docker-bgp/minigraph_facts.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
+#!/usr/bin/env python
```
Comment (qiluo-msft) [inline dockers/docker-bgp/minigraph_facts.py]: There are some concerns about
1. duplicate the minigraph parsers in every containers.
2. runtime overhead to generate config files on each bootup

The requirement is as:
1. there is no change of minigraph.xml after deployment
2. so, it is possible to create config files on host, copy to docker containers after they are just created. only the first bootup will be slower (if possible, we can run this step before reboot)

File: dockers/docker-bgp/Dockerfile [hunk 5]
```diff
@@ -3,11 +3,15 @@ FROM docker-base
...
```
Comment (qiluo-msft) [inline dockers/docker-bgp/Dockerfile]: >etc [](start = 54, length = 3)

/etc is not a good place for scripts.

File: dockers/docker-bgp/minigraph_facts.py (hunk 4 above)
Comment (taoyl-ms) [inline dockers/docker-bgp/minigraph_facts.py]: For 1, we are considering move those config-related scripts into docker-base after they are stabilized.

For 2, Guohan has some concern on potential changes of config file schema update container version update, thus we want to keep the capability of generating config file in the container itself. 

Maybe Guohan could give more comments on this?

Comment (lguohan) [inline dockers/docker-bgp/minigraph_facts.py]: for 1, we can move it do the docker-base and stablized there.

for 2, configuration generate should be part of the docker container's job. For example, if we switch quagga to gobgp, the configuration template for them is going to be very different, where should be store the config template? I feel should be stored in the container.

File: dockers/docker-bgp/render_config.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    #parser.add_argument("-v", "--variable-file", help="yaml file that contains addtional variables")
+    args = parser.parse_args()
+
+    minigraph = args.minigraph
+    template_file = args.template
+    
+    data = parse_xml(minigraph)
+
+    env = jinja2.Environment(loader=jinja2.FileSystemLoader('./'), trim_blocks=True)
+    env.filters['ipv4'] = is_ipv4
```
Comment (qiluo-msft) [inline dockers/docker-bgp/render_config.py]: >filters [](start = 8, length = 7)

How many kinds of filters will we use? If many, we should prevent consider implementation and maintenance cost. One option is to reuse ansible implementation.

File: dockers/docker-bgp/config.sh [hunk 7]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+cd /etc/swss
+
+awk '{system("python render_config.py -m /etc/swss/minigraph.xml "$1 ">"$2" && chown "$3" "$2" && chmod "$4" "$2)}' template_list
```
Comment (qiluo-msft) [inline dockers/docker-bgp/config.sh]: Is this command too complex to read?

File: dockers/docker-bgp/render_config.py (hunk 6 above)
Comment (taoyl-ms) [inline dockers/docker-bgp/render_config.py]: ansible implementation is of GPLv3. There might be copyright concerns.

File: dockers/docker-bgp/template_list [hunk 8]
```diff
@@ -0,0 +1,4 @@
+bgpd.conf.j2  /etc/quagga/bgpd.conf    quagga:quaggavty  0644
+zebra.conf.j2 /etc/quagga/zebra.conf   root:root         0644
+isolate.j2    /usr/sbin/bgp-isolate    root:root         0755
+unisolate.j2  /usr/sbin/bgp-unisolate  root:root         0755
```
Comment (qiluo-msft) [inline dockers/docker-bgp/template_list]: >nisolate.j2 [](start = 1, length = 11)

suggest implement this feature as:
1. make fake config files in Dockerfile, set the correct ownership, groupship and mode. I guess the vanilla ones are already good.
2. copy file and maintain the target file attribute. ref: [http://serverfault.com/questions/273949/cp-not-want-to-overwrite-permissions/273956](http://serverfault.com/questions/273949/cp-not-want-to-overwrite-permissions/273956)

File: dockers/docker-bgp/minigraph_facts.py (hunk 4 above)
Comment (stcheng) [inline dockers/docker-bgp/minigraph_facts.py]: minigraph_facts.py could be a python library that could be used by other config generators

File: dockers/docker-bgp/config.sh (hunk 7 above)
Comment (lguohan) [inline dockers/docker-bgp/config.sh]: agree, use variable name would make it more readable.

**Comment by lguohan** [comment]
where is this file? https://github.com/Azure/sonic-mgmt/blob/master/ansible/roles/sonicv2/templates/quagga/daemons

File: dockers/docker-bgp/Dockerfile [hunk 9]
```diff
@@ -3,11 +3,15 @@ FROM docker-base
 COPY deps/quagga_*.deb /deps/
 RUN dpkg_apt() { [ -f $1 ] && { dpkg -i $1 || apt-get -y install -f; } || return 1; } &&        \
     dpkg_apt /deps/quagga_*.deb &&                                                              \
+    apt-get update &&                                                                           \
+    apt-get -y install -f python-jinja2 python-netaddr python-ipaddr python-lxml &&             \
     apt-get clean -y && apt-get autoclean -y && apt-get autoremove -y &&                        \
     rm -rf /deps
 
 COPY daemons /etc/quagga/
```
Comment (lguohan) [inline dockers/docker-bgp/Dockerfile]: >daemons [](start = 5, length = 7)

where is this file?

Comment (taoyl-ms) [inline dockers/docker-bgp/Dockerfile]: it's already there. somebody add it in a previous version. 
and, it's actually not a template

File: dockers/docker-bgp/Dockerfile (hunk 1 above)
Comment (qiluo-msft) [inline dockers/docker-bgp/Dockerfile]: Discussed with @lguohan @taoyl-ms, not to move to docker-base. Instead, package it and install in all the related docker images.

File: dockers/docker-bgp/minigraph_facts.py (hunk 4 above)
Comment (qiluo-msft) [inline dockers/docker-bgp/minigraph_facts.py]: I agree the configuration generation code should be inside each container.

**Pull request closed without merging**